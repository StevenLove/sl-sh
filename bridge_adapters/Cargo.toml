[package]
name = "bridge_adapters"
version = "0.1.0"
edition = "2021"

[dependencies]
slvm = { workspace = true }
compile_state = { workspace = true }
bridge_types = { workspace = true }
