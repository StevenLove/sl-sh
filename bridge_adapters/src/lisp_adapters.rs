//! TODO PC ISSUE #8 need explanation for the emulation for TryFrom/TryInto/AsRef/AsMut
//! My notes:
//! #. To convert a slosh &Value to an owned type implement `impl SlFrom<&Value> for OwnedType`,
//!     this allows rust native functions annotated with the bridge macro to receive normal
//!     rust types.
//! #. To convert a slosh &Value to a reference type implement `impl SlAsRef<&Value> for RefType`.
//! #. To convert a slosh &Value to a mutable reference type implement `impl SlAsMut<&Value> for MutRefType`.
//! #. To convert some rust type back to a value that the rust native function
//!     annotated by the bridge macro returns implement `impl SlFrom<&Value> for RustType`.
//!     TODO PC ISSUE #8 blanket impl so impl `SlFrom<Value>` works, and taking a ref isn't required?
//! #. To avoid allocations when converting a slosh &Value back to a rust type that was mutated
//!     don't return anything. If it is necessary for the API to return some value,
//!     TODO PC ISSUE #7 annotated or lifetime?
//!
//!
//! ## rosetta stone for bridge macros
//! Rust Type                   | Slosh Type & Traits   <br>&emsp; <br> S -> R Convert Slosh -> Rust <br> &emsp; - Occurs when coercing slush arguments to the parameter types in the signature of the annotated Rust function. <br> R -> S Convert Rust -> Slosh <br> &emsp; - Occurs when coercing some returned Rust type to a Slosh type. |
//! ----------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------|
//! [`String`]                  | [Value]`::String`           |
//!                             |                             | S -> R
//!                             |                             |     &emsp;- [`SlInto`] [`String`] for `&`[`Value`]
//!                             |                             | R -> S
//!                             |                             |     &emsp;- [`SlFrom`] `&`[`Value`] for [`String`]
//!                             |                             |
//! `&`[`String`]               | [`Value`]`::String`         |
//!                             |                             | S -> R
//!                             |                             |     &emsp;- [`SlInto`] `&`[`String`] for `&`[`Value`]
//!                             |                             | R -> S
//!                             |                             |     &emsp;- take [`String`]
//!                             |                             |     &emsp;* uses Clone unless TODO PC ISSUE #7 the extant value problem
//!                             |                             |
//! `&mut `[`String`]           | [`Value`]`::String`         |
//!                             |                             | S -> R
//!                             |                             |     &emsp;- [`SlAsMut`] [`String`] for `&`[`Value`]
//!                             |                             | R -> S
//!                             |                             |     &emsp;- take `&mut `[`String`]
//!                             |                             |     &emsp;* uses Clone unless TODO PC ISSUE #7 the extant value problem
//!                             |                             |
//! `&`[`str`]                  | [`Value`]`::String` / [`Value`]`::StringConst` |
//!                             |                             | S -> R
//!                             |                             |     &emsp;- [`SlAsRef`] [`str`] for `&`[`Value`]
//!                             |                             | R -> S
//!                             |                             |     &emsp;- [`SlFrom`] for [`Value`]
//!                             |                             |     &emsp;* uses Clone unless TODO PC ISSUE #7 - the extant value problem
//!                             |                             |     &emsp;- TODO PC ISSUE #7 adjacent is it even possible to call vm.alloc_string_ro on something that was *newly* created in the current fcn and returned as a RO value OR should that be made as a custom type so the user can declare their intent.
//!                             |                             |
//! [`char`]                    | [`Value`]`::CodePoint`      |
//!                             |                             | S -> R
//!                             |                             |     &emsp;- [`SlInto`] [`char`] for `&`[`Value`]
//!                             |                             | R -> S
//!                             |                             |     &emsp;- [`SlFrom`] `&`[`Value`] for [`char`]
//!                             |                             |
//! [SloshChar]               |  [`Value`]`::CharClusterLong` / [`Value`]`::CharCluster` / [`Value`]`::CodePoint` |
//!                             |                             | S -> R
//!                             |                             |     &emsp;- [`SlIntoRef`] [`SloshChar`] for `&`[`Value`]
//!                             |                             | R -> S
//!                             |                             |     &emsp;- [`SlFromRef`] `&`[`Value`] for [`SloshChar`]
//!                             |                             |
//! [`LooseString`]             | [`Value`]`::String` / [`Value`]`::CodePoint` / [`Value`]`::CharCluster` / [`Value`]`::CharClusterLong` / [`Value`]`::Symbol` / [`Value`]`::Keyword` / [`Value`]`::StringConst` |
//!                             |                             | S -> R
//!                             |                             |     &emsp;- [`SlIntoRef`] [`LooseString`] for `&`[`Value`]
//!                             |                             | R -> S
//!                             |                             |     &emsp;* Note: Always does an allocation and returns a [`Value`]`::String` type.
//!                             |                             |     &emsp;- [`SlFromRef`] `&`[`Value`] for [`LooseString`]
//!                             |                             |
//! [`bool`]                    | [`Value`]::True / [`Value`]::False / [`Value`]::Nil |
//!                             |                             | S -> R
//!                             |                             |     &emsp;- [`Into`] [`bool`] for `&`[`Value`]
//!                             |                             | R -> S
//!                             |                             |     &emsp;- [`SlFrom`] `&`[`Value`] for [`primitives`]
//!                             |                             |
//!                             |                             |
//!                             |                             |
//!                             |                             |
//!                             |                             |
//!                             |                             |
//!  [`Value`]::Byte(u8),
//!  [`Value`]::Int([u8; 7]), // Store a 7 byte int (i56...).
//!  [`Value`]::Float(F56),
//!
//!  [`Value`]::Pair(Handle),
//!  [`Value`]::List(Handle, u16),
//!  [`Value`]::Vector(Handle),
//!  [`Value`]::Map(Handle),
//!
//!  [`Value`]::Symbol(Interned),
//!  [`Value`]::Keyword(Interned),
//!  [`Value`]::Special(Interned), // Intended for symbols that are compiled.
//!  [`Value`]::Builtin(u32),
//!  [`Value`]::Undefined,
//!  [`Value`]::Nil,
//!
//!  [`Value`]::Bytes(Handle),
//!  [`Value`]::Lambda(Handle),
//!  [`Value`]::Closure(Handle),
//!  [`Value`]::Continuation(Handle),
//!  [`Value`]::CallFrame(Handle),
//!  [`Value`]::Error(Handle),

#[cfg(doc)]
use bridge_types::{LooseString, SloshChar};

use compile_state::state::SloshVm;

use slvm::{VMResult, Value};
pub mod numbers;
pub mod primitives;
pub mod string_char;

//TODO should this only ever be implemented for Value?
//  e.g. impl SlFrom<RustType> for Value per RustType?
// that way when we take a mutable reference it's when
// we're going back to slosh when we might need a mutable
// reference to allocate memory for stuff in the call.
// this is in contrast to SlFromRef<'a, &'a Value> for RustType
// where we want to be able to have an immutable reference to
// the vm to borrow n things from the vm (really the heap immutably).

// TODO use marker trait here T: BridgedType, to make it clear
// which rust types have been implemented for slosh.
pub trait SlFrom<T>: Sized {
    /// Converts to this type from the input type.
    fn sl_from(value: T, vm: &mut SloshVm) -> VMResult<Self>;
}

pub trait SlInto<T>: Sized {
    /// Converts this type into the (usually inferred) input type.
    fn sl_into(self, vm: &mut SloshVm) -> VMResult<T>;
}

impl<T, U> SlInto<U> for T
where
    U: SlFrom<T>,
{
    fn sl_into(self, vm: &mut SloshVm) -> VMResult<U> {
        U::sl_from(self, vm)
    }
}

pub trait SlFromRef<'a, T>: Sized
where
    Self: 'a,
{
    /// Converts to this type from the input type.
    fn sl_from_ref(value: T, vm: &'a SloshVm) -> VMResult<Self>;
}

pub trait SlIntoRef<'a, T>: Sized
where
    T: 'a,
{
    /// Converts to this type from the input type.
    fn sl_into_ref(self, vm: &'a SloshVm) -> VMResult<T>;
}

impl<'a, T, U> SlIntoRef<'a, U> for T
where
    U: SlFromRef<'a, T>,
    U: 'a,
{
    fn sl_into_ref(self, vm: &'a SloshVm) -> VMResult<U> {
        U::sl_from_ref(self, vm)
    }
}

pub trait SlAsRef<'a, T: ?Sized> {
    /// Converts this type into a shared reference of the (usually inferred) input type.
    fn sl_as_ref(&self, vm: &'a SloshVm) -> VMResult<&'a T>;
}

// SlAsRef lifts over &
impl<'a, T: ?Sized, U: ?Sized> SlAsRef<'a, U> for &'a T
where
    T: SlAsRef<'a, U>,
{
    #[inline]
    fn sl_as_ref(&self, vm: &'a SloshVm) -> VMResult<&'a U> {
        <T as SlAsRef<'a, U>>::sl_as_ref(*self, vm)
    }
}

// SlAsRef lifts over &mut
impl<'a, T: ?Sized, U: ?Sized> SlAsRef<'a, U> for &'a mut T
where
    T: SlAsRef<'a, U>,
{
    #[inline]
    fn sl_as_ref(&self, vm: &'a SloshVm) -> VMResult<&'a U> {
        <T as SlAsRef<'a, U>>::sl_as_ref(*self, vm)
    }
}

//TODO consider removing me to flush out which values shouldn't
// need to implement SlFrom and where the macro is doing that.
// even worth keeping around to check at compile time that
// nothing implements SlFrom<&Value> for T.
// could I?
// /// PANICS do not implement!
// ///
// impl<T> SlFrom<&Value> for T {
//     fn sl_from(_value: &Value, _vm: &mut SloshVm) -> VMResult<Self> {
//         unimplemented!()
//     }
// }
// impl<T> SlFrom<&Value> for T
// where
//     T: for<'a> SlFromRef<'a, &'a Value>,
// {
//     fn sl_from(value: &Value, vm: &mut SloshVm) -> VMResult<Self> {
//         Self::sl_from_ref(value, vm)
//     }
// }

pub trait SlAsMut<'a, T: ?Sized> {
    /// Converts this type into a mutable reference of the (usually inferred) input type.
    fn sl_as_mut(&mut self, vm: &'a mut SloshVm) -> VMResult<&'a mut T>;
}

// SlAsMut lifts over &mut
impl<'a, T: ?Sized, U: ?Sized> SlAsMut<'a, U> for &'a mut T
where
    T: SlAsMut<'a, U>,
{
    #[inline]
    fn sl_as_mut(&mut self, vm: &'a mut SloshVm) -> VMResult<&'a mut U> {
        (*self).sl_as_mut(vm)
    }
}

impl SlFrom<Value> for Value {
    fn sl_from(value: Value, _vm: &mut SloshVm) -> VMResult<Self> {
        Ok(value)
    }
}
