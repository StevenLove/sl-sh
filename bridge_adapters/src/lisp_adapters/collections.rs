use crate::lisp_adapters::{SlFromRef, SlFromRefMut};
use bridge_types::ErrorStrings;
use compile_state::state::SloshVm;
use slvm::{VMError, VMResult, Value, ValueType};
use std::collections::HashMap;

impl<'a> SlFromRef<'a, Value> for &'a HashMap<Value, Value> {
    fn sl_from_ref(value: Value, vm: &'a SloshVm) -> VMResult<Self> {
        match value {
            Value::Map(h) => Ok(vm.get_map(h)),
            _ => Err(VMError::new_conversion(
                ErrorStrings::fix_me_mismatched_type(
                    <&'static str>::from(ValueType::Map),
                    value.display_type(vm),
                ),
            )),
        }
    }
}

impl<'a> SlFromRefMut<'a, Value> for &'a mut HashMap<Value, Value> {
    fn sl_from_ref_mut(value: Value, vm: &'a mut SloshVm) -> VMResult<Self> {
        match value {
            Value::Map(h) => Ok(vm.get_map_mut(h)?),
            _ => Err(VMError::new_conversion(
                ErrorStrings::fix_me_mismatched_type(
                    <&'static str>::from(ValueType::Map),
                    value.display_type(vm),
                ),
            )),
        }
    }
}
