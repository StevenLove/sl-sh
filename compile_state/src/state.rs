use std::cell::RefCell;
use std::collections::HashMap;
use std::rc::Rc;

use slvm::chunk::*;
use slvm::interner::*;
use slvm::value::*;
use slvm::vm::*;

#[derive(Clone, Debug)]
pub struct SymbolsInt {
    pub syms: HashMap<Interned, usize>,
    count: usize,
}

impl SymbolsInt {
    pub fn add_sym(&mut self, sym: Interned) {
        self.syms.insert(sym, self.count);
        self.count += 1;
    }
}

// symbol name, idx/reg for scope, idx/reg for outer scope
type Captures = Rc<RefCell<Vec<(Interned, usize, usize)>>>;

#[derive(Clone, Debug)]
pub struct Symbols {
    pub data: Rc<RefCell<SymbolsInt>>,
    outer: Option<Rc<RefCell<Symbols>>>,
    pub captures: Captures,
}

impl Symbols {
    pub fn with_outer(outer: Option<Rc<RefCell<Symbols>>>) -> Symbols {
        let data = Rc::new(RefCell::new(SymbolsInt {
            syms: HashMap::new(),
            count: 1,
        }));
        Symbols {
            data,
            outer,
            captures: Rc::new(RefCell::new(Vec::new())),
        }
    }

    pub fn with_let(source: Rc<RefCell<Symbols>>) -> Symbols {
        let data = Rc::new(RefCell::new(SymbolsInt {
            syms: HashMap::new(),
            count: source.borrow().data.borrow().count,
        }));
        {
            let mut datad = data.borrow_mut();
            for (key, val) in source.borrow().data.borrow().syms.iter() {
                datad.syms.insert(*key, *val);
            }
        }
        Symbols {
            data,
            outer: source.borrow().outer.clone(),
            captures: source.borrow().captures.clone(),
        }
    }

    pub fn is_empty(&self) -> bool {
        self.data.borrow().syms.is_empty()
    }

    pub fn regs_count(&self) -> usize {
        self.data.borrow().count
    }

    pub fn contains_symbol(&self, key: Interned) -> bool {
        self.data.borrow().syms.contains_key(&key)
    }

    pub fn can_capture(&self, key: Interned) -> bool {
        let mut loop_outer = self.outer.clone();
        while let Some(outer) = loop_outer {
            let outer = outer.borrow();
            if outer.contains_symbol(key) {
                return true;
            }
            loop_outer = outer.outer.clone();
        }
        false
    }

    pub fn get_capture_binding(&self, key: Interned) -> Option<usize> {
        for cap in &*self.captures.borrow() {
            if cap.0 == key {
                return Some(cap.2);
            }
        }
        None
    }

    pub fn get(&self, key: Interned) -> Option<usize> {
        self.data.borrow().syms.get(&key).copied()
    }

    pub fn clear(&mut self) {
        self.data.borrow_mut().syms.clear();
    }

    pub fn reserve_reg(&mut self) -> usize {
        let mut data = self.data.borrow_mut();
        let count = data.count;
        data.count += 1;
        count
    }

    pub fn insert(&mut self, key: Interned) -> usize {
        let mut data = self.data.borrow_mut();
        let count = data.count;
        data.syms.insert(key, count);
        data.count += 1;
        count
    }

    pub fn insert_reserved(&mut self, key: Interned, register: usize) {
        let mut data = self.data.borrow_mut();
        data.syms.insert(key, register);
    }

    pub fn insert_capture(&self, vm: &mut SloshVm, key: Interned) -> Option<usize> {
        let data_d = self.data.borrow();
        if let Some(idx) = data_d.syms.get(&key) {
            Some(*idx)
        } else {
            if let Some(outer) = &self.outer {
                drop(data_d);
                // Also capture in outer lexical scope or bad things can happen.
                if let Some(outer_idx) = outer.borrow().insert_capture(vm, key) {
                    let mut data = self.data.borrow_mut();
                    let count = data.count;
                    data.syms.insert(key, count);
                    data.count += 1;
                    self.captures.borrow_mut().push((key, count, outer_idx));
                    return Some(count);
                }
            }
            None
        }
    }

    pub fn len_captures(&self) -> usize {
        self.captures.borrow().len()
    }
}

pub struct Specials {
    pub def: Interned,
    pub set: Interned,
    pub do_: Interned,
    pub fn_: Interned,
    pub mac_: Interned,
    pub if_: Interned,
    pub add: Interned,
    pub sub: Interned,
    pub mul: Interned,
    pub div: Interned,
    pub inc: Interned,
    pub dec: Interned,
    pub list: Interned,
    pub list_append: Interned,
    pub cons: Interned,
    pub car: Interned,
    pub cdr: Interned,
    pub xar: Interned,
    pub xdr: Interned,
    pub vec: Interned,
    pub make_vec: Interned,
    pub vec_pop: Interned,
    pub vec_push: Interned,
    pub vec_nth: Interned,
    pub vec_set: Interned,
    pub quote: Interned,
    pub backquote: Interned,
    pub recur: Interned,
    pub this_fn: Interned,
    pub numeq: Interned,
    pub numneq: Interned,
    pub numlt: Interned,
    pub numlte: Interned,
    pub numgt: Interned,
    pub numgte: Interned,
    pub eq: Interned,
    pub equal: Interned,
    pub type_: Interned,
    pub not: Interned,
    pub and: Interned,
    pub or: Interned,
    pub err: Interned,
    pub vec_len: Interned,
    pub vec_clr: Interned,
    pub str_: Interned,
    pub let_: Interned,
    pub call_cc: Interned,
    pub defer: Interned,
    pub on_error: Interned,
    pub while_: Interned,
    pub doc_string: Interned,

    pub rest: Interned,
    pub optional: Interned,
    pub scratch: Interned,
}

impl Specials {
    pub fn new(vm: &mut SloshVm) -> Self {
        Self {
            def: vm.intern_static("def"),
            set: vm.intern_static("set!"),
            do_: vm.intern_static("do"),
            fn_: vm.intern_static("fn"),
            mac_: vm.intern_static("macro"),
            if_: vm.intern_static("if"),
            add: vm.intern_static("+"),
            sub: vm.intern_static("-"),
            mul: vm.intern_static("*"),
            div: vm.intern_static("/"),
            inc: vm.intern_static("inc!"),
            dec: vm.intern_static("dec!"),
            list: vm.intern_static("list"),
            list_append: vm.intern_static("list-append"),
            cons: vm.intern_static("cons"),
            car: vm.intern_static("car"),
            cdr: vm.intern_static("cdr"),
            xar: vm.intern_static("xar!"),
            xdr: vm.intern_static("xdr!"),
            vec: vm.intern_static("vec"),
            make_vec: vm.intern_static("make-vec"),
            vec_push: vm.intern_static("vec-push!"),
            vec_pop: vm.intern_static("vec-pop!"),
            vec_nth: vm.intern_static("vec-nth"),
            vec_set: vm.intern_static("vec-set!"),
            quote: vm.intern_static("quote"),
            backquote: vm.intern_static("back-quote"),
            recur: vm.intern_static("recur"),
            this_fn: vm.intern_static("this-fn"),
            numeq: vm.intern_static("="),
            numneq: vm.intern_static("/="),
            numlt: vm.intern_static("<"),
            numlte: vm.intern_static("<="),
            numgt: vm.intern_static(">"),
            numgte: vm.intern_static(">="),
            eq: vm.intern_static("eq?"),
            equal: vm.intern_static("equal?"),
            type_: vm.intern_static("type"),
            not: vm.intern_static("not"),
            and: vm.intern_static("and"),
            or: vm.intern_static("or"),
            err: vm.intern_static("err"),
            vec_len: vm.intern_static("vec-len"),
            vec_clr: vm.intern_static("vec-clear!"),
            str_: vm.intern_static("str"),
            let_: vm.intern_static("let"),
            call_cc: vm.intern_static("call/cc"),
            defer: vm.intern_static("defer"),
            on_error: vm.intern_static("on-error"),
            while_: vm.intern_static("while"),
            doc_string: vm.intern_static("doc-string"),

            rest: vm.intern_static("&"),
            optional: vm.intern_static("%"),
            scratch: vm.intern_static("[SCRATCH]"),
        }
    }
}

pub struct CompileState {
    pub symbols: Rc<RefCell<Symbols>>,
    pub constants: HashMap<Value, usize>,
    pub chunk: Chunk,
    pub max_regs: usize,
    pub tail: bool,
    pub defers: usize,
    pub doc_string: Option<Value>,
}

impl Default for CompileState {
    fn default() -> Self {
        Self::new()
    }
}

impl CompileState {
    pub fn new() -> Self {
        CompileState {
            symbols: Rc::new(RefCell::new(Symbols::with_outer(None))),
            constants: HashMap::new(),
            chunk: Chunk::new("no_file", 1),
            max_regs: 0,
            tail: false,
            defers: 0,
            doc_string: None,
        }
    }

    pub fn new_state(
        file_name: &'static str,
        first_line: u32,
        outer: Option<Rc<RefCell<Symbols>>>,
    ) -> Self {
        let symbols = Rc::new(RefCell::new(Symbols::with_outer(outer)));
        CompileState {
            symbols,
            constants: HashMap::new(),
            chunk: Chunk::new(file_name, first_line),
            max_regs: 0,
            tail: false,
            defers: 0,
            doc_string: None,
        }
    }

    pub fn reserved_regs(&self) -> usize {
        self.symbols.borrow().regs_count()
    }

    pub fn get_symbol(&self, sym: Interned) -> Option<usize> {
        self.symbols.borrow().data.borrow().syms.get(&sym).copied()
    }

    pub fn add_constant(&mut self, exp: Value) -> usize {
        if let Some(i) = self.constants.get(&exp) {
            *i
        } else {
            let const_i = self.chunk.add_constant(exp);
            self.constants.insert(exp, const_i);
            const_i
        }
    }
}

pub struct CompileEnvironment {
    use_line: bool,
    line: u32,
    specials: Option<Specials>,
    global_map: HashMap<Interned, usize>,
}

impl Default for CompileEnvironment {
    fn default() -> Self {
        Self::new()
    }
}

impl CompileEnvironment {
    pub fn new() -> Self {
        Self {
            use_line: true,
            line: 1,
            specials: None,
            global_map: HashMap::new(),
        }
    }
}

pub type SloshVm = GVm<CompileEnvironment>;

pub trait SloshVmTrait {
    fn set_line_val(&mut self, state: &mut CompileState, val: Value);
    fn get_reserve_global(&mut self, symbol: Interned) -> u32;
    fn set_named_global(&mut self, string: &str, value: Value) -> u32;
    fn set_global_builtin(&mut self, string: &str, func: CallFuncSig<CompileEnvironment>) -> u32;
    fn dump_globals(&self);
    fn own_line(&self) -> Option<u32>;
    fn line_num(&self) -> u32;
    fn specials(&self) -> &Specials;
    fn global_intern_slot(&self, symbol: Interned) -> Option<u32>;
}

pub fn new_slosh_vm() -> SloshVm {
    let temp_env = CompileEnvironment::new();
    let mut vm = GVm::new_with_env(temp_env);
    let specials = Specials::new(&mut vm);
    vm.env_mut().specials = Some(specials);
    vm
}

impl SloshVmTrait for SloshVm {
    fn set_line_val(&mut self, state: &mut CompileState, val: Value) {
        if let (Some(Value::UInt32(dline)), Some(Value::StringConst(file_intern))) = (
            self.get_heap_property(val, "dbg-line"),
            self.get_heap_property(val, "dbg-file"),
        ) {
            let file_name = self.get_interned(file_intern);
            if file_name == state.chunk.file_name && dline as u32 > self.env().line {
                self.env_mut().line = dline as u32;
            }
        }
    }

    fn get_reserve_global(&mut self, symbol: Interned) -> u32 {
        if let Some(idx) = self.env().global_map.get(&symbol) {
            *idx as u32
        } else {
            let idx = self.reserve_global();
            self.env_mut().global_map.insert(symbol, idx as usize);
            idx
        }
    }

    fn set_named_global(&mut self, string: &str, value: Value) -> u32 {
        let sym = self.intern(string);
        let slot = self.get_reserve_global(sym);
        self.set_global(slot, value);
        slot
    }

    fn set_global_builtin(&mut self, string: &str, func: CallFuncSig<CompileEnvironment>) -> u32 {
        let f_val = self.add_builtin(func);
        self.set_named_global(string, f_val)
    }

    fn dump_globals(&self) {
        println!("GLOBALS:");
        let mut ordered_keys = Vec::with_capacity(self.env().global_map.len());
        ordered_keys.resize(self.env().global_map.len(), "");
        for (k, v) in self.env().global_map.iter() {
            ordered_keys[*v] = self.get_interned(*k);
        }
        for (i, k) in ordered_keys.iter().enumerate() {
            println!(
                "({:#010x})/{}: {}",
                i,
                *k,
                self.get_global(i as u32).display_value(self)
            );
        }
        println!();
    }

    fn own_line(&self) -> Option<u32> {
        if self.env().use_line {
            Some(self.env().line)
        } else {
            None
        }
    }

    fn line_num(&self) -> u32 {
        if self.env().use_line {
            self.env().line
        } else {
            0
        }
    }

    fn specials(&self) -> &Specials {
        if self.env().specials.is_none() {}
        self.env().specials.as_ref().unwrap()
    }

    fn global_intern_slot(&self, symbol: Interned) -> Option<u32> {
        self.env()
            .global_map
            .get(&symbol)
            .copied()
            .map(|i| i as u32)
    }
}
