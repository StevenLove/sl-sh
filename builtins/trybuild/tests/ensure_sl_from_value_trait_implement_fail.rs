use bridge_adapters::lisp_adapters::{SlAsRef, SlInto};
use compile_state::state::new_slosh_vm;

pub fn main() {}
fn do_not_implement_sl_from_to_convert_from_value_types_to_rust_types() {
    let mut vm = new_slosh_vm();
    let dest = vm.alloc_string("XXX".to_string());
    let dest2 = vm.alloc_string("YYY".to_string());

    let x: String = (&dest).sl_into(&mut vm).unwrap();
    let x: f64 = (&dest).sl_into(&mut vm).unwrap();
    let x: i64 = (&dest).sl_into(&mut vm).unwrap();
    let x: usize = (&dest).sl_into(&mut vm).unwrap();
    let x: f32 = (&dest).sl_into(&mut vm).unwrap();
    let x: &str = (&dest).sl_into(&mut vm).unwrap();
    let x: bool = (&dest).sl_into(&mut vm).unwrap();
}
