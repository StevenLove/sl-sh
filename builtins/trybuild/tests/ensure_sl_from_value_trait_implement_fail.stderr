error[E0277]: the trait bound `String: SlFrom<&slvm::value::Value>` is not satisfied
  --> trybuild/tests/ensure_sl_from_value_trait_implement_fail.rs:11:29
   |
11 |     let x: String = (&dest).sl_into(&mut vm).unwrap();
   |                             ^^^^^^^ the trait `SlFrom<&slvm::value::Value>` is not implemented for `String`
   |
   = help: the following other types implement trait `SlFrom<T>`:
             <slvm::value::Value as SlFrom<bool>>
             <slvm::value::Value as SlFrom<char>>
             <slvm::value::Value as SlFrom<i32>>
             <slvm::value::Value as SlFrom<u32>>
             <slvm::value::Value as SlFrom<f64>>
             <slvm::value::Value as SlFrom<bridge_types::SloshChar<'a>>>
             <slvm::value::Value as SlFrom<Cow<'a, str>>>
             <slvm::value::Value as SlFrom<slvm::value::Value>>
           and $N others
   = note: required for `&slvm::value::Value` to implement `SlInto<String>`

error[E0277]: the trait bound `f64: SlFrom<&slvm::value::Value>` is not satisfied
  --> trybuild/tests/ensure_sl_from_value_trait_implement_fail.rs:12:26
   |
12 |     let x: f64 = (&dest).sl_into(&mut vm).unwrap();
   |                          ^^^^^^^ the trait `SlFrom<&slvm::value::Value>` is not implemented for `f64`
   |
   = help: the following other types implement trait `SlFrom<T>`:
             <slvm::value::Value as SlFrom<bool>>
             <slvm::value::Value as SlFrom<char>>
             <slvm::value::Value as SlFrom<i32>>
             <slvm::value::Value as SlFrom<u32>>
             <slvm::value::Value as SlFrom<f64>>
             <slvm::value::Value as SlFrom<bridge_types::SloshChar<'a>>>
             <slvm::value::Value as SlFrom<Cow<'a, str>>>
             <slvm::value::Value as SlFrom<slvm::value::Value>>
           and $N others
   = note: required for `&slvm::value::Value` to implement `SlInto<f64>`

error[E0277]: the trait bound `i64: SlFrom<&slvm::value::Value>` is not satisfied
  --> trybuild/tests/ensure_sl_from_value_trait_implement_fail.rs:13:26
   |
13 |     let x: i64 = (&dest).sl_into(&mut vm).unwrap();
   |                          ^^^^^^^ the trait `SlFrom<&slvm::value::Value>` is not implemented for `i64`
   |
   = help: the following other types implement trait `SlFrom<T>`:
             <slvm::value::Value as SlFrom<bool>>
             <slvm::value::Value as SlFrom<char>>
             <slvm::value::Value as SlFrom<i32>>
             <slvm::value::Value as SlFrom<u32>>
             <slvm::value::Value as SlFrom<f64>>
             <slvm::value::Value as SlFrom<bridge_types::SloshChar<'a>>>
             <slvm::value::Value as SlFrom<Cow<'a, str>>>
             <slvm::value::Value as SlFrom<slvm::value::Value>>
           and $N others
   = note: required for `&slvm::value::Value` to implement `SlInto<i64>`

error[E0277]: the trait bound `usize: SlFrom<&slvm::value::Value>` is not satisfied
  --> trybuild/tests/ensure_sl_from_value_trait_implement_fail.rs:14:28
   |
14 |     let x: usize = (&dest).sl_into(&mut vm).unwrap();
   |                            ^^^^^^^ the trait `SlFrom<&slvm::value::Value>` is not implemented for `usize`
   |
   = help: the following other types implement trait `SlFrom<T>`:
             <slvm::value::Value as SlFrom<bool>>
             <slvm::value::Value as SlFrom<char>>
             <slvm::value::Value as SlFrom<i32>>
             <slvm::value::Value as SlFrom<u32>>
             <slvm::value::Value as SlFrom<f64>>
             <slvm::value::Value as SlFrom<bridge_types::SloshChar<'a>>>
             <slvm::value::Value as SlFrom<Cow<'a, str>>>
             <slvm::value::Value as SlFrom<slvm::value::Value>>
           and $N others
   = note: required for `&slvm::value::Value` to implement `SlInto<usize>`

error[E0277]: the trait bound `f32: SlFrom<&slvm::value::Value>` is not satisfied
  --> trybuild/tests/ensure_sl_from_value_trait_implement_fail.rs:15:26
   |
15 |     let x: f32 = (&dest).sl_into(&mut vm).unwrap();
   |                          ^^^^^^^ the trait `SlFrom<&slvm::value::Value>` is not implemented for `f32`
   |
   = help: the following other types implement trait `SlFrom<T>`:
             <slvm::value::Value as SlFrom<bool>>
             <slvm::value::Value as SlFrom<char>>
             <slvm::value::Value as SlFrom<i32>>
             <slvm::value::Value as SlFrom<u32>>
             <slvm::value::Value as SlFrom<f64>>
             <slvm::value::Value as SlFrom<bridge_types::SloshChar<'a>>>
             <slvm::value::Value as SlFrom<Cow<'a, str>>>
             <slvm::value::Value as SlFrom<slvm::value::Value>>
           and $N others
   = note: required for `&slvm::value::Value` to implement `SlInto<f32>`

error[E0277]: the trait bound `&str: SlFrom<&slvm::value::Value>` is not satisfied
  --> trybuild/tests/ensure_sl_from_value_trait_implement_fail.rs:16:27
   |
16 |     let x: &str = (&dest).sl_into(&mut vm).unwrap();
   |                           ^^^^^^^ the trait `SlFrom<&slvm::value::Value>` is not implemented for `&str`
   |
   = help: the following other types implement trait `SlFrom<T>`:
             <slvm::value::Value as SlFrom<bool>>
             <slvm::value::Value as SlFrom<char>>
             <slvm::value::Value as SlFrom<i32>>
             <slvm::value::Value as SlFrom<u32>>
             <slvm::value::Value as SlFrom<f64>>
             <slvm::value::Value as SlFrom<bridge_types::SloshChar<'a>>>
             <slvm::value::Value as SlFrom<Cow<'a, str>>>
             <slvm::value::Value as SlFrom<slvm::value::Value>>
           and $N others
   = note: required for `&slvm::value::Value` to implement `SlInto<&str>`

error[E0277]: the trait bound `bool: SlFrom<&slvm::value::Value>` is not satisfied
  --> trybuild/tests/ensure_sl_from_value_trait_implement_fail.rs:17:27
   |
17 |     let x: bool = (&dest).sl_into(&mut vm).unwrap();
   |                           ^^^^^^^ the trait `SlFrom<&slvm::value::Value>` is not implemented for `bool`
   |
   = help: the following other types implement trait `SlFrom<T>`:
             <slvm::value::Value as SlFrom<bool>>
             <slvm::value::Value as SlFrom<char>>
             <slvm::value::Value as SlFrom<i32>>
             <slvm::value::Value as SlFrom<u32>>
             <slvm::value::Value as SlFrom<f64>>
             <slvm::value::Value as SlFrom<bridge_types::SloshChar<'a>>>
             <slvm::value::Value as SlFrom<Cow<'a, str>>>
             <slvm::value::Value as SlFrom<slvm::value::Value>>
           and $N others
   = note: required for `&slvm::value::Value` to implement `SlInto<bool>`
